### DISCUSSION TITLE
Confusion about Class Blade Components

### DISCUSSION THREAD

**Discussion by nathangross** [tighten/jigsaw#590] [Q&A] [answered]
Hopefully this is a quick one: where would I put this `Layout.php` file ref'd in #541 ?

```
<?php

namespace Components;

use Illuminate\View\Component;
use Illuminate\Container\Container;

class Layout extends Component
{
    public $page;

    public function __construct()
    {
        $this->page = Container::getInstance()->make('pageData');
    }

    public function render()
    {        
        return Container::getInstance()->make('view')->make('_components.layout');
    }
}
```

I too am trying to utilize blade components for my layout files and hoping to not have to pass the $page var each time. I came across #492 and #541 but I wasn't exactly sure where that php file should go. 

Thanks!

**Comment by bakerkretzmar** [comment, marked as answer]
You can actually put the PHP files wherever you want as long as you namespace and autoload them properly. Jigsaw will discover any components that are in the global `Components\` namespace, so just autoload the directory that your component's PHP files are in. See #507.

If you put them in `app/Components` for example, add this to your `composer.json`:

```json
"autoload": {
  "psr-4": {
    "Components\\": "app/Components/"
  }
}
```

and `composer dump` and you should be all set.

**Reply by nathangross** [reply]
For some reason I still can't get this to work. 

`Notice: Undefined variable: page`

In the example above, the `app/Components` directory structure would be in the root of the project (not in the `source` directory), correct? 

Also, assuming `app/Components` is in the root of the project, would I need to specify `make('_components.layout')` or `make('source._components.layout')`?

I realize this is more of a composer/psr-4 question than a Jigsaw question...

Thanks!

**Reply by bakerkretzmar** [reply]
@nathangross looks like my example in #541 is wrong 🤦🏻 it should be this:

```diff
-         $this->page = Container::getInstance()->make('pageData');
+         $this->page = Container::getInstance()->make('pageData')->page;
```

**Reply by nathangross** [reply]
Got it! Thank you!

Another thing that was throwing me off was that the **namespace and class directory needs to match up with the component directory**. It took some messing around to figure this out and I think it makes sense now, but wasn't clear to me before.
